content/name-based-xml.md

```diff
@@ -0,0 +1,421 @@
+---
+layout: sip
+title: SIP-NN - Name Based XML Literals
+vote-status: pending
+permalink: /sips/:title.html
+redirect_from: /sips/pending/name-based-xml.html
+---
+
+**By: Yang, Bo**
+
+## History
+
+| Date          | Version       |
+|---------------|---------------|
+| Oct 24th 2019 | Initial Draft |
+
+## Introduction
+
+Name-based `for` comprehension has been proven success in Scala language design. A `for` / `yield` expression will be converted to higher-order function calls to `flatMap` , `map` and `withFilter` methods, no matter which type signatures they are. The `for` comprehension can be used for either `Option` or `List` , even when `List` has an additional implicit `CanBuildFrom` parameter. Third-party libraries like Scalaz and Cats also provides `Ops` to allow monadic data types in `for` comprehension.
+
+[Name-based pattern matching](https://dotty.epfl.ch/docs/reference/changed-features/pattern-matching.html) is introduced by Dotty. It is greatly simplified the implementation compared to Scala 2. In addition, specific symbols in Scala library ( `Option` , `Seq` ) are decoupled from the Scala compiler.
+
+Considering the success of the above name-based syntactic sugars, in order to decouple `scala-xml` library from Scala compiler, name-based XML literal is an obvious approach.
+
+## Motivating Examples
+
+### Examples
+
+Given an XML literal `<div title="my-title">line1<br/>line2</div>`, at the parser phase, the compiler should internally convert it to an AST equivalent to the following code:
+
+``` scala
+xml.literal(
+  xml.elements.div(
+    xml.attributes.title(xml.values.`my-title`),
+    xml.texts.line1,
```

Comment:
A related discussion https://github.com/ScalablyTyped/Converter/discussions/343

Reply:
I don't think the erasure issue is a problem here. For `text`, I believe a single result type will be enough anyway. And for `value`, I would expect a unique method like
```scala
def value(s: String): Value[s.type] = ???
```
to be sufficient. At call site, the expected type `Value["checked"]` would be enough.

If *really* that's not good enough, there's always `@targetName`.

Reply:
`Value["checked"]` would be good. I tried the approach in Scala 2 but there are some issues that I did not remember preventing it from working. I guess it should work in Scala 3, because the literal type support is more solid.

Reply:
ScalablyTyped is using some weird trick to mimic literal types https://github.com/ScalablyTyped/Converter/blob/dfaf906684b3dbe74aff26457b4e900e8be8bb77/docs/encoding.md?plain=1#L251

I forgot the exact issue, but Scala 2 compiler tends to use the widen type everywhere possible, and resulted in some compile error that could be avoid if a parameter is inferred as the literal type.

Reply:
Another reason is that when I was writing this proposal I was trying to support Scala 2.12 or even 2.11, where literal types are not available.

Reply:
I can update this proposal to use string types instead. Maybe also for attribute names and tag names to avoid name clash with `toString`?